src/sentry/snuba/metrics/metric_extraction.py

```diff
@@ -0,0 +1,258 @@
+
+    Tags values can be static if defined through `value` or dynamically queried
+    from the payload if defined through `field`. These two options are mutually
+    exclusive, behavior is undefined if both are specified.
+    """
+
+    key: str
+    field: NotRequired[str]
+    value: NotRequired[str]
+    condition: NotRequired[RuleCondition]
+
+
+class MetricSpec(TypedDict):
+    """
+    Specification for a metric to extract from some data.
+
+    The metric type is given as part of the MRI (metric reference identifier)
+    which must follow the form: `<type>:<namespace>/<name>@<unit>`.
+
+    How the metric's value is obtained depends on the metric type:
+     - Counter metrics are a special case, since the default product counters do
+       not count any specific field but rather the occurrence of the event. As
+       such, there is no value expression, and the field is set to `None`.
+       Semantics of specifying remain undefined at this point.
+     - Distribution metrics require a numeric value.
+     - Set metrics require a string value, which is then emitted into the set as
+       unique value. Insertion of numbers and other types is undefined.
+    """
+
+    category: Literal["transaction"]
+    mri: str
+    field: NotRequired[Optional[str]]
+    condition: NotRequired[RuleCondition]
+    tags: NotRequired[Sequence[TagSpec]]
+
+
+def convert_alert_to_metric(snuba_query: SnubaQuery) -> Optional[MetricSpec]:
+    if snuba_query.dataset != Dataset.PerformanceMetrics.value:
+        return None
+
+    try:
+        spec = OndemandMetricSpec.parse(snuba_query.aggregate, snuba_query.query)
+        if not spec:
+            return None
+
+        return {
+            "category": DataCategory.TRANSACTION.api_name(),
+            "mri": spec.mri,
+            "field": spec.field,
+            "condition": spec.condition(),
+            "tags": [{"key": QUERY_HASH_KEY, "value": spec.query_hash()}],
+        }
+    except Exception as e:
+        logger.error(e, exc_info=True)
+        return None
+
+
+@dataclass(frozen=True)
+class OndemandMetricSpec:
+    _field: str
+    _query: str
+
+    field: Optional[str]
+    metric_type: str
+    mri: str
+    op: MetricOperationType
+
+    @classmethod
+    def check(cls, field: str, query: str) -> bool:
+        """Returns ``True`` if a metrics query requires an on-demand metric."""
+        # TODO?: self.dataset == Dataset.PerformanceMetrics.value and
+        return "transaction.duration" in query
+
+    @classmethod
+    def parse(cls, field: str, query: str) -> Optional["OndemandMetricSpec"]:
+        if not cls.check(field, query):
+            return None
+
+        relay_field, metric_type, op = _extract_field_info(field)
+
+        return cls(
+            _field=field,
+            _query=query,
+            field=relay_field,
+            metric_type=metric_type,
+            mri=f"{metric_type}:{CUSTOM_ALERT_METRIC_NAME}@none",
+            op=op,
+        )
```

Comment:
Having static/class methods for such operations, especially considering that they are mostly used together and share the same set of params is a bit counterintuitive (it can work fine, just not the best API to interact with). In addition, they call some logic which is tied to parsing but it's all over the file as functions.

A more self-contained solution could be:
```python
class OnDemandMetricGenerator:

	def __init__(field, query):
		...

	def check():
		...
	
	def parse():
		...

	# All stateful private methods responsible of parsing.
```

What do you think?

Reply:
Is that class you're suggesting a factory, or does it internally hold parse state? If so, how would you indicate that parsing doesn't work and make usage intuitive?

Explaining the original rationale: Initially, `check` and `parse` were internal functions returning tuples, but now they are part of a public signature. We could also make them plain functions again, where only one of them returns an `OnDemandMetricSpec` if that is more idiomatic than a class method. I'm probably applying patterns from other languages here.

The public interface needs to be concise, as that defines the contract of on-demand metric generation and usage. I would suggest to keep that entire contract in one class (or set of functions), as _defining_ and _usage_ of such a metric are intrinsically coupled together.

As for which functions are called internally, that's an implementation detail of this module. The structure isn't ideal yet, though only once we're adding more capabilities (count_if, failure_rate, more mapping functions), we see which transformations are needed.